Cython/Compiler/Nodes.py

```diff
@@ -3899,25 +3920,27 @@ def generate_keyword_unpacking_code(self, min_positional_args, max_positional_ar
         code.putln('switch (pos_args) {')
         if self.star_arg:
             code.putln('default:')
-        for i in range(max_positional_args-1, -1, -1):
+
+        for i in range(max_positional_args-1, num_required_posonly_args-1, -1):
             code.put('case %2d: ' % (i+1))
-            if i+1 > num_required_posonly_args:
+            code.putln("values[%d] = PyTuple_GET_ITEM(%s, %d);" % (
+                i, Naming.args_cname, i))
+            code.putln('CYTHON_FALLTHROUGH;')
+        if num_required_posonly_args > 0:
+            code.put('case %2d: ' % num_required_posonly_args)
+            for i in range(num_required_posonly_args-1, -1, -1):
                 code.putln("values[%d] = PyTuple_GET_ITEM(%s, %d);" % (
                     i, Naming.args_cname, i))
-                code.putln('CYTHON_FALLTHROUGH;')
-            elif i+1 == num_required_posonly_args:
-                for j in range(num_required_posonly_args-1, -1, -1):
-                    code.putln("values[%d] = PyTuple_GET_ITEM(%s, %d);" % (
-                        j, Naming.args_cname, j))
-                code.putln('break;')
-            else:
-                code.putln('CYTHON_FALLTHROUGH;')
+            code.putln('break;')
+        for i in range(num_required_posonly_args-2, -1, -1):
+            code.put('case %2d: ' % (i+1))
+            code.putln('CYTHON_FALLTHROUGH;')
```

Comment:
This part looks a bit redundant with the `default:` case below – but I guess it's safer to just always generate it.

Reply:
Yeah, when there is no *arg these cases are redundant (though not redundant when there is a *arg).  I thought about dealing with this case differently, but wasn't sure if it was worth adding yet another case to the already complicated logic.  Happy to add this case if you think it's a good trade-off, for now I've left it as it is.

Reply:
I think it's ok to let the C compiler deal with it.